docs/CONTROLFLOW.md

```diff
@@ -0,0 +1,343 @@
+3) Applies [block splitting](#block-splitting)
+4) Generates [junk jumps](#junk-jumps)
+5) Applies [control flow flattening](#control-flow-flattening)
+6) Converts go/ssa back into go/ast
+
+### Example usage
+
+```go
+
+// Obfuscate with defaults parameters
+//garble:controlflow
+func main() {
+	println("Hello world!")
+}
+```
+
+### Parameter explanation
+
+> Unlike other garble features (which just work), we recommend that you understand how parameters affect control flow obfuscation and which caveats exist.
+
+> Code snippets below without name obfuscation, for better readability.
+
+#### Block splitting
+
+Param: `block_splits` (default: `0`)
+
+> Warning: this param affects resulting binary only when used in combination with [flattening](#control-flow-flattening)
+
+
+Block splitting splits the largest block into 2 parts of random size, this is done `block_splits` times at most. If there is no more suitable block to split (number of ssa instructions in block is less than 3), splitting stops.
+
+This param is very useful if your code has few branches (`if`, `switch` etc.).
+
+Input:
+```go
+package main
+
+// Note that the block_splits value is very large, so code blocks are split into the smallest possible blocks.
+//garble:controlflow flatten_passes=0 junk_jumps=0 block_splits=1024
+func main() {
+	println("1")
+	println("2")
+	println("3")
+	println("4")
+	println("5")
+}
+```
+
+Result:
+
+```go
+func main() {
+	{
+		println("1")
+		goto _s2a_l3
+	}
+_s2a_l1:
+	{
+		println("3")
+		goto _s2a_l4
+	}
+_s2a_l2:
+	{
+		println("5")
+		return
+	}
+_s2a_l3:
+	{
+		println("2")
+		goto _s2a_l1
+	}
+_s2a_l4:
+	{
+		println("4")
+		goto _s2a_l2
+	}
+}
+ 
+```
+
+
+#### Junk jumps
+
+Param: `junk_jumps` (default: `0`)
+
+> Warning: this param affects resulting binary only when used in combination with [flattening](#control-flow-flattening)
+
+Junk jumps adds jumps to random blocks `junk_jumps` times, these inserted jumps can also form a chain. This param is useful for linearly increasing the functions complexity.
+
+Input:
+```go
+//garble:controlflow flatten_passes=0 junk_jumps=5 block_splits=0
+func main() {
+	if true {
+		println("hello world")
+	}
+}
+```
+
+Result: 
+
+```go
+func main() {
+	{
+		if true {
+			goto _s2a_l4
+		} else {
+			goto _s2a_l2
+		}
+	}
+_s2a_l1:
+	{
+		println("hello world")
+		goto _s2a_l7
+	}
+_s2a_l2:
+	{
+		return
+	}
+_s2a_l3:
+	{
+		goto _s2a_l2
+	}
+_s2a_l4:
+	{
+		goto _s2a_l6
+	}
+_s2a_l5:
+	{
+		goto _s2a_l3
+	}
+_s2a_l6:
+	{
+		goto _s2a_l1
+	}
+_s2a_l7:
+	{
+		goto _s2a_l5
+	}
+}
+```
+
+#### Control flow flattening
+
+Param: `flatten_passes` (default: `1`)
+
+
+This parameter completely [flattens the control flow](https://github.com/obfuscator-llvm/obfuscator/wiki/Control-Flow-Flattening) `flatten_passes` times, which makes analysing the logic of the function very difficult
+This is the most important parameter without which the other parameters have no effect on the resulting binary.
```

Comment:
Does garble fail if the other parameters are used but this one is not? I hope it does, otherwise this feels like a bit of a sharp edge.

Reply:
it can be useful for "exporting" source code or debugging. I think a warning in log would be enough